### ISSUE TITLE
Inferring NumPy array type when using `as_tensor`

### ISSUE THREAD

**Issue by sumanthratna** [pytorch/pytorch#40568]
## 🚀 Feature
Let the `device` keyword argument of [`torch.as_tensor`](https://pytorch.org/docs/master/generated/torch.as_tensor.html#torch.as_tensor) be either a `np.dtype` or `torch.dtype`. 

## Motivation

Suppose I have two numpy arrays with different types and I want to convert one of them to a torch tensor with the type of the other array. 

According to https://discuss.pytorch.org/t/converting-a-numpy-dtype-to-torch-dtype/52279/2, there's no convenient way to convert a numpy tensor to a torch tensor. 

## Pitch

```python
import numpy as np
import torch


# currently raises the following:
# TypeError: as_tensor(): argument 'dtype' must be 
# torch.dtype, not numpy.dtype
mytensor = torch.tensor(48., dtype=np.float32)
```

## Alternatives

- make [this API](https://github.com/pytorch/pytorch/blob/29881c7f02ed7cb263a602f73a84089a79571c8e/torch/csrc/utils/tensor_numpy.cpp#L196-L220) public
- let the `device` keyword argument of [`torch.Tensor.to`](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.to) be either a `np.dtype` or `torch.dtype`

## Additional context

Similar issue: #541

**Comment by vadimkantorov** [comment]
"device" keyword? typo? did you mean "dtype" keyword?

**Comment by sumanthratna** [comment]
> "device" keyword? typo? did you mean "dtype" keyword?

Yikes, that was a typo. I've edited the original issue.

**Comment by mruberry** [comment]
Thanks for the suggestion, @sumanthratna! This would be cool. Although if we did it for this function I'd like to see PyTorch do it universally, and I'm concerned that might require checking the value passed to the dtype kwarg (which could have a performance impact). I wonder if either (1) there's a simpler way to accomplish this or (2) we should just encourage people to copy/paste the dict we created for our own conversions: https://github.com/pytorch/pytorch/blob/ac79c874cefee2f8bc1605eed9a924d80c0b3542/torch/testing/_internal/common_utils.py#L349.

**Comment by vadimkantorov** [comment]
@sumanthratna I think that if you convert bumpy dtype to string like ‘int8’, it will already work

**Comment by sumanthratna** [comment]
> Although if we did it for this function I'd like to see PyTorch do it universally

For sure! I do see why you're concerned about the performance impact; maybe we can wrap the dict lookup in an `if not isinstance(dtype, torch.dtypes)`, where `torch.dtypes` is a tuple of available dtypes. This might save us a little time in the case that a torch dtype is passed in, but we'll need to benchmark because isinstance might be slower. 

> we should just encourage people to copy/paste the dict we created for our own conversions

This is a very minor complaint but I feel like torch should have this functionality, even if it's just in a method called `torch.dtype.to_numpy(self)` (note that I have no idea how the `torch.dtype` namespace works). I guess I'm *okay* with `torch.as_tensor(npy_array, dtype=torch.dtype.to_numpy(np.int8))`. Personally, it makes for slightly annoying code when I have to copy the same dict into all of my torch projects. 

> I think that if you convert bumpy dtype to string like ‘int8’, it will already work
```python
import numpy as np
import torch


a = np.array(4.)
b = np.array(5.)
print(torch.as_tensor(a, dtype=b.dtype.name))
```
raises: `TypeError: as_tensor(): argument 'dtype' must be torch.dtype, not str`. 

I would be okay with `torch.as_tensor(npy_array, dtype=np.int8.name)`. Accepting strings or torch dtypes (as opposed to numpy and torch dtypes) might be faster because I think we could use 
```python
if isinstance(dtype, str):
    dtype = getattr(torch, dtype)
```

**Comment by vadimkantorov** [comment]
I agree, supporting string string dtypes would be a good idea, also for compat with NumPy:
```python
import torch
import numpy

numpy.array([3, 3], dtype = 'int32')
# array([3, 3], dtype=int32)

torch.tensor([3, 3], dtype = 'int32')
# TypeError: tensor(): argument 'dtype' must be torch.dtype, not str
```
